platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+- `Chapter` may only be a child of `Router`
+- Its only children may be `Page`s
+- A `Chapter` must have children
+
+**Responsibilities:**
+- Render the chapter title and progress bar
+- Render the child `Page`s
+
+##### `Page` Component
+**Purpose:**
+A `Page` is used to define the path for a page and render its form contents and
+the normal page chrome.
+
+**Usage requirements:**
+A `Page` may be the child of either `Router` directly or `Chapter`.
+
+**Responsibilities:**
+- If no `overrideFormik` prop is passed
+  - Render a [`<Formik>`](https://formik.org/docs/api/formik) component and pass
+    it some default props for `initialValues`, `validate`, and `onSubmit`
+  - If the save-in-progress React context is available, `Page` will use it to
+    auto-save the form
+- Render the children, passing them the extra props for form state
+  - For an example usage, see [below](#functionality-array-pages)
+  - **Caveat:** This means when the children of `Page` are a collection of
+    `<Field>`s (or similar) instead of the functional component in the example
+    below, they'll all have "extra" props for the form context
+    - This is the simplest way to support both simple and complex form pages,
+      but may be subject to change later if we can figure out a better way
+- Render the form navigation buttons "Back" and "Continue"
+- Prevent navigation if there are validation errors
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
+
+**Usage requirements:**
+- `Route` may only be a child of `Router`
+
+**Why not just use `Page`?**
+The `Chapter` and `Page` components are meant to be used to define pages where
+the user may enter data. These pages are sequential and have back and continue
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <FieldArray
+      name="marriages"
+      render={arrayHelpers => (
+        <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+      )}
+    />
+  )}
+</Page>
+```
+
+For array pages to work we need a couple core things:
+- The path for the page
+- The data for the current item in the array
+
+To get these pieces, `Page` will accept the `path` and `pathParams` props.
+
+`path` will be the URL within the application and must (for array pages) contain
+URL parameters (term pending). URL parameters must be the only token between
```

Comment:
Would "slug" get the point across?

Reply:
Isn't a slug a different concept? I thought that was the last portion of the URL's path.

Reply:
I think a slug is the entire portion of a URL after the domain.